File: src/main/java/space/davids_digital/neurobots/gui/GuiApp.java
package space.davids_digital.neurobots.gui;

import space.davids_digital.neurobots.geom.DoublePoint;
import space.davids_digital.neurobots.model.Creature;
import space.davids_digital.neurobots.model.NeuralNetwork;
import space.davids_digital.neurobots.model.Wall;
import space.davids_digital.neurobots.model.World;

import javax.swing.*;
import java.awt.*;

public class GuiApp {
    public static void main(String[] args) {
        new GuiApp();
    }

    private WorldViewer worldViewer;

    private World world;
    private JFrame frame;

    private GuiApp() {
        initWorld();
        initGui();
        initUpdating();
        initRendering();
    }

    private void initGui() {
        frame = new JFrame("Neurobots");
        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        frame.setMinimumSize(new Dimension(1200, 800));
        frame.setVisible(true);

        worldViewer = new WorldViewer(world);
        frame.getContentPane().add(worldViewer);
        worldViewer.grabFocus();
    }

    private void initWorld() {
        world = new World(1200, 800);
        world.getWalls().add(new Wall(new DoublePoint(0, 0), new DoublePoint(0, world.getHeight())));
        world.getWalls().add(new Wall(new DoublePoint(0, world.getHeight()), new DoublePoint(world.getWidth(), world.getHeight())));
        world.getWalls().add(new Wall(new DoublePoint(world.getWidth(), world.getHeight()), new DoublePoint(world.getWidth(), 0)));
        world.getWalls().add(new Wall(new DoublePoint(world.getWidth(), 0), new DoublePoint(0, 0)));
        for (int i = 0; i < 30; i++)
            world.getCreatures().add(new Creature(
                    new NeuralNetwork(21, 1, 11, 3, -0.25, 0.25),
                    Color.green,
                    new DoublePoint(100 + 10*i, 40),
                    500, 10, 0, 3, 12, 12, 30, Math.toRadians(45)
            ));
    }

    private void initUpdating() {
        var ref = new Object() {
            double lastUpdate = System.currentTimeMillis();
        };
        Timer timer = new Timer(10, event -> {
            double delta = System.currentTimeMillis() - ref.lastUpdate;
            ref.lastUpdate = System.currentTimeMillis();
            world.update(delta);
            worldViewer.update(delta);
        });
        timer.setRepeats(true);
        timer.start();
    }

    private void initRendering() {
        Timer timer = new Timer(10, event -> frame.repaint());
        timer.setRepeats(true);
        timer.start();
    }
}

File: src/main/java/space/davids_digital/neurobots/gui/GuiApp.kt
package space.davids_digital.neurobots.gui

import space.davids_digital.neurobots.game.Config
import space.davids_digital.neurobots.geom.DoublePoint
import space.davids_digital.neurobots.world.*
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Dimension
import java.lang.Math.random
import javax.swing.JFrame
import javax.swing.Timer
import javax.swing.UIManager
import javax.swing.WindowConstants
import kotlin.math.min
import kotlin.math.sin

class GuiApp private constructor() {
    private lateinit var worldViewer: WorldViewer
    private lateinit var world: World
    private lateinit var frame: JFrame
    private var config = Config()

    init {

        initWorld()
        initGui()
        initUpdating()
        initRendering()
    }

    private fun initGui() {
        UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName())
        frame = JFrame("Neurobots")
        frame.defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        frame.size = Dimension(1200, 800)
        worldViewer = WorldViewer(world, config)
        frame.layout = BorderLayout()
        frame.add(worldViewer, BorderLayout.CENTER)
        frame.add(ControlPanel(world), BorderLayout.EAST)
        worldViewer.grabFocus()
        frame.isVisible = true
    }

    private fun initWorld() {
        world = World(2000, 2000, 0.005, 0.2, 1.0)
        val w = world.width - 1
        val h = world.height - 1
        world += Wall(0, 0, 0, h)
        world += Wall(0, h, w, h)
        world += Wall(w, h, w, 0)
        world += Wall(w, 0, 0, 0)
        world += Wall(w/2, h/2 - h/5, w/2, h/2 + h/5)
        world += Wall(w/2 - w/5, h/2, w/2 + w/5, h/2)
        world += FoodSpawner(DoublePoint(0, 0), DoublePoint(world.width, world.height), 20.0, 10.0, 100)
        repeat(10) {
            world += CreatureSpawner(
                DoublePoint(random()*world.width, random()*world.height),
                Color(random().toFloat(), random().toFloat(), random().toFloat()),
                2,
                8,
                1000.0,
                7,
                50.0, 200.0,
                50.0, 100.0,
                50.0,
                Math.toRadians(75.0),
                4
            )
        }
    }

    private fun initUpdating() {
        var lastUpdate = System.currentTimeMillis().toDouble()
        val timer = Timer(10) {
            val delta = min(System.currentTimeMillis() - lastUpdate, 30.0)
            lastUpdate = System.currentTimeMillis().toDouble()
            world.update(delta)
            worldViewer.update(delta)
        }
        timer.isRepeats = true
        timer.start()
    }

    private fun initRendering() {
        val timer = Timer(10) { frame.repaint() }
        timer.isRepeats = true
        timer.start()
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            GuiApp()
        }
    }
}